File: src/com/nadmm/airports/wx/Metar.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

package com.nadmm.airports.wx;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.EnumSet;

public final class Metar implements Serializable {

    private static final long serialVersionUID = 1L;

    public enum Flags {
        Corrected {
            @Override
            public String toString() {
                return "Corrected METAR";
            }
        },
        AutoStation {
            @Override
            public String toString() {
                return "Automated station";
            }
        },
        AutoReport {
            @Override
            public String toString() {
                return "Automated observation with no human augmentation";
            }
        },
        MaintenanceIndicatorOn {
            @Override
            public String toString() {
                return "Station needs maintenance";
            }
        },
        PresentWeatherSensorOff {
            @Override
            public String toString() {
                return "Present weather sensor is not operating";
            }
        },
        LightningSensorOff {
            @Override
            public String toString() {
                return "Lightning detection sensor is not operating";
            }
        },
        RainSensorOff {
            @Override
            public String toString() {
                return "Rain sensor is not operating";
            }
        },
        FreezingRainSensorOff {
            @Override
            public String toString() {
                return "Freezing rain sensor is not operating";
            }
        };
    }

    public boolean isValid;
    public String stationId;
    public String rawText;
    public long observationTime;
    public long fetchTime;
    public float tempCelsius;
    public float dewpointCelsius;
    public int windDirDegrees;
    public int windSpeedKnots;
    public int windGustKnots;
    public int windPeakKnots;
    public float visibilitySM;
    public float altimeterHg;
    public float seaLevelPressureMb;
    public ArrayList<WxSymbol> wxList;
    public ArrayList<SkyCondition> skyConditions;
    public String flightCategory;
    public float pressureTend3HrMb;
    public float maxTemp6HrCentigrade;
    public float minTemp6HrCentigrade;
    public float maxTemp24HrCentigrade;
    public float minTemp24HrCentigrade;
    public float precipInches;
    public float precip3HrInches;
    public float precip6HrInches;
    public float precip24HrInches;
    public float snowInches;
    public int vertVisibilityFeet;
    public String metarType;
    public float stationElevationMeters;

    public EnumSet<Flags> flags;
    public boolean presrr;
    public boolean presfr;
    public boolean snincr;
    public boolean wshft;
    public boolean fropa;

    public Metar() {
        isValid = false;
        tempCelsius = Float.MAX_VALUE;
        dewpointCelsius = Float.MAX_VALUE;
        windDirDegrees = Integer.MAX_VALUE;
        windSpeedKnots = Integer.MAX_VALUE;
        windGustKnots = Integer.MAX_VALUE;
        windPeakKnots = Integer.MAX_VALUE;
        visibilitySM = Float.MAX_VALUE;
        altimeterHg = Float.MAX_VALUE;
        seaLevelPressureMb = Float.MAX_VALUE;
        pressureTend3HrMb = Float.MAX_VALUE;
        maxTemp6HrCentigrade = Float.MAX_VALUE;
        minTemp6HrCentigrade = Float.MAX_VALUE;
        maxTemp24HrCentigrade = Float.MAX_VALUE;
        minTemp24HrCentigrade = Float.MAX_VALUE;
        precipInches = Float.MAX_VALUE;
        precip3HrInches = Float.MAX_VALUE;
        precip6HrInches = Float.MAX_VALUE;
        precip24HrInches = Float.MAX_VALUE;
        snowInches = Float.MAX_VALUE;
        vertVisibilityFeet = Integer.MAX_VALUE;
        stationElevationMeters = Float.MAX_VALUE;

        wxList = new ArrayList<WxSymbol>();
        skyConditions = new ArrayList<SkyCondition>();

        flags = EnumSet.noneOf( Flags.class );
        presrr = false;
        presfr = false;
        snincr = false;
        wshft = false;
        fropa = false;
    }

    public void computeFlightCategory() {
        int ceiling = 12000;
        for ( SkyCondition sky : skyConditions ) {
            // Ceiling is defined as the lowest layer aloft reported as broken or overcast;
            // or the vertical visibility into an indefinite ceiling
            if ( sky.name().equals( "BKN" ) 
                    || sky.name().equals( "OVC" ) 
                    || sky.name().equals( "OVX" ) ) {
                ceiling =sky.getCloudBase();
                break;
            }
        }
        if ( ceiling < 500 || visibilitySM < 1.0 ) {
            flightCategory = "LIFR";
        } else if ( ceiling < 1000 || visibilitySM < 3.0 ) {
            flightCategory = "IFR";
        } else if ( ceiling <= 3000 || visibilitySM <= 5.0 ) {
            flightCategory = "MVFR";
        } else {
            flightCategory = "VFR";
        }
    }

    public void setMissingFields() {
        if ( flightCategory == null ) {
            computeFlightCategory();
        }
        if ( vertVisibilityFeet < Integer.MAX_VALUE ) {
            // Check to see if we have an OVX layer, if not add it
            boolean found = false;
            for ( SkyCondition sky : skyConditions ) {
                if ( sky.name().equals( "OVX" ) ) {
                    found = true;
                    break;
                }
            }
            if ( !found ) {
                skyConditions.add( SkyCondition.create( "OVX", 0 ) );
            }
        }
        if ( skyConditions.isEmpty() ) {
            // Sky condition is not available in the METAR
            skyConditions.add( SkyCondition.create( "SKM", 0 ) );
        }
    }

}

File: src/main/java/com/nadmm/airports/wx/Metar.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011-2023 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.wx

import android.os.Parcelable
import com.nadmm.airports.utils.WxUtils
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.Serializable

enum class MetarFlag(var description: String) {
    CORRECTED("Corrected METAR"),
    AUTOMATED_STATION("Automated station"),
    MAINTENANCE_INDICATOR_ON("Station needs maintenance"),
    PRESENT_WEATHER_SENSOR_OFF("Present weather sensor is not operating"),
    LIGHTNING_SENSOR_OFF("Lightning detection sensor is not operating"),
    RAIN_SENSOR_OFF("Rain sensor is not operating"),
    FREEZING_RAIN_SENSOR_OFF("Freezing rain sensor is not operating"),
    NO_SIGNAL("No signal from station");

    override fun toString(): String = description
}

enum class WeatherPhenomenon(var description: String) {
    PRESSURE_RISING_RAPIDLY("Pressure rising rapidly"),
    PRESSURE_FALLING_RAPIDLY("Pressure falling rapidly"),
    SNOW_INCREASING_RAPIDLY("Snow increasing rapidly"),
    WIND_SHIFT("Wind shift"),
    FRONTAL_PASSAGE("Frontal passage"),
    LIGHTNING("Lightning detected");

    override fun toString(): String = description
}

@Parcelize
@Serializable
data class Metar(
    var stationId: String? = null,
    var fetchTime: Long = 0L,
    var isValid: Boolean = false,
    var rawText: String? = null,
    var observationTime: Long = 0L,
    var tempCelsius: Float = Float.MAX_VALUE,
    var dewpointCelsius: Float = Float.MAX_VALUE,
    var windDirDegrees: Int = Int.MAX_VALUE,
    var windSpeedKnots: Int = Int.MAX_VALUE,
    var windGustKnots: Int = Int.MAX_VALUE,
    var windPeakKnots: Int = Int.MAX_VALUE,
    var visibilitySM: Float = Float.MAX_VALUE,
    var altimeterHg: Float = Float.MAX_VALUE,
    var seaLevelPressureMb: Float = Float.MAX_VALUE,
    var wxList: MutableList<WxSymbol> = mutableListOf(),
    var skyConditions: MutableList<SkyCondition> = mutableListOf(),
    var flightCategory: String = WxUtils.FLIGHT_CATEGORY_UNKN,
    var pressureTend3HrMb: Float = Float.MAX_VALUE,
    var maxTempCelsiusLast6Hours: Float = Float.MAX_VALUE,
    var minTempCelsiusLast6Hours: Float = Float.MAX_VALUE,
    var maxTempCelsiusLast24Hours: Float = Float.MAX_VALUE,
    var minTempCelsiusLast24Hours: Float = Float.MAX_VALUE,
    var precipInches: Float = Float.MAX_VALUE,
    var precip3HrInches: Float = Float.MAX_VALUE,
    var precip6HrInches: Float = Float.MAX_VALUE,
    var precip24HrInches: Float = Float.MAX_VALUE,
    var snowInches: Float = Float.MAX_VALUE,
    var vertVisibilityFeet: Int = Int.MAX_VALUE,
    var metarType: String? = null,
    var stationElevationMeters: Float = Float.MAX_VALUE,
    var flags: MutableSet<MetarFlag> = mutableSetOf(),
    var weatherPhenomena: MutableSet<WeatherPhenomenon> = mutableSetOf()
) : Parcelable
